chromadb/api/models/CollectionCommon.py

```diff
@@ -450,52 +426,76 @@ def _update_model_after_modify_success(
+    def _process_add_request(
         self,
         ids: OneOrMany[ID],
-        embeddings: Optional[  # type: ignore[type-arg]
+        embeddings: Optional[
             Union[
                 OneOrMany[Embedding],
                 OneOrMany[np.ndarray],
             ]
         ],
         metadatas: Optional[OneOrMany[Metadata]],
         documents: Optional[OneOrMany[Document]],
-        images: Optional[OneOrMany[Image]],
-        uris: Optional[OneOrMany[URI]],
-    ) -> Tuple[
-        IDs,
-        Embeddings,
-        Optional[Metadatas],
-        Optional[Documents],
-        Optional[URIs],
-    ]:
-        (
-            ids,
-            embeddings,
-            metadatas,
-            documents,
-            images,
-            uris,
-        ) = self._validate_embedding_set(
+        images: Optional[OneOrMany[Image]] = None,
+        uris: Optional[OneOrMany[URI]] = None,
+    ) -> EmbeddingSet:
+        unpacked_embedding_set = self._unpack_embedding_set(
             ids,
             embeddings,
             metadatas,
             documents,
             images,
             uris,
+        )
+
+        normalized_embeddings = (
+            self._normalize_embeddings(unpacked_embedding_set["embeddings"])
+            if unpacked_embedding_set["embeddings"] is not None
+            else None
+        )
+
+        self._validate_embedding_set(
+            unpacked_embedding_set["ids"],
+            normalized_embeddings,
+            unpacked_embedding_set["metadatas"],
+            unpacked_embedding_set["documents"],
+            unpacked_embedding_set["images"],
+            unpacked_embedding_set["uris"],
             require_embeddings_or_data=False,
         )
 
+        prepared_embeddings = self._prepare_embeddings(
+            normalized_embeddings,
+            unpacked_embedding_set["documents"],
+            unpacked_embedding_set["images"],
+            unpacked_embedding_set["uris"],
+        )
+
+        return {
+            "ids": unpacked_embedding_set["ids"],
+            "embeddings": prepared_embeddings,
+            "metadatas": unpacked_embedding_set["metadatas"],
+            "documents": unpacked_embedding_set["documents"],
+            "images": unpacked_embedding_set["images"],
+            "uris": unpacked_embedding_set["uris"],
+        }
+
+    def _prepare_update_request(
+        self,
+        embeddings: Optional[Embeddings],
+        documents: Optional[Documents],
+        images: Optional[Images],
+    ) -> Embeddings:
         if embeddings is None:
             if documents is not None:
                 embeddings = self._embed(input=documents)
             elif images is not None:
                 embeddings = self._embed(input=images)
 
-        return ids, cast(Embeddings, embeddings), metadatas, documents, uris
+        return cast(Embeddings, embeddings)
 
-    def _validate_and_prepare_upsert_request(
+    def _process_update_request(
```

Comment:
`process_update` and `process_upsert` appear to have identical logic.

Reply:
I think if we make `update` and `upsert` work with URIs, the logic would be identical for all of these.

Reply:
consolidated into one